Kernel/Devices/DeviceManagement.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2021, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/HashTable.h>
+#include <AK/Singleton.h>
+#include <Kernel/Devices/DeviceManagement.h>
+#include <Kernel/FileSystem/InodeMetadata.h>
+#include <Kernel/Sections.h>
+
+namespace Kernel {
+
+static Singleton<DeviceManagement> s_the;
+
+UNMAP_AFTER_INIT DeviceManagement::DeviceManagement()
+{
+}
+UNMAP_AFTER_INIT void DeviceManagement::initialize()
+{
+    s_the.ensure_instance();
+}
+
+UNMAP_AFTER_INIT void DeviceManagement::attach_null_device(Device const& device)
+{
+    VERIFY(device.major() == 1);
+    VERIFY(device.minor() == 3);
+    m_null_device = device;
+}
+
+DeviceManagement& DeviceManagement::the()
+{
+    return *s_the;
+}
+
+Device* DeviceManagement::get_device(unsigned major, unsigned minor)
+{
+    return m_devices.with_exclusive([&](auto& map) -> Device* {
+        auto it = map.find(encoded_device(major, minor));
+        if (it == map.end())
+            return nullptr;
+        return it->value;
+    });
+}
+
+void DeviceManagement::before_removal(Badge<Device>, Device& device)
+{
+    u32 device_id = encoded_device(device.major(), device.minor());
+    m_devices.with_exclusive([&](auto& map) -> void {
+        VERIFY(map.contains(device_id));
+        map.remove(encoded_device(device.major(), device.minor()));
+    });
+}
+
+void DeviceManagement::after_inserting(Badge<Device>, Device& device)
+{
+    u32 device_id = encoded_device(device.major(), device.minor());
+    m_devices.with_exclusive([&](auto& map) -> void {
+        auto it = map.find(device_id);
+        if (it != map.end()) {
+            dbgln("Already registered {},{}: {}", device.major(), device.minor(), it->value->class_name());
+        }
+        VERIFY(!map.contains(device_id));
```

Comment:
these two checks are equivalent aren't they? So instead of duplicating the logic we can just do:
```suggestion
        if (map.contains(device_id)) {
            dbgln("Already registered {},{}: {}", device.major(), device.minor(), it->value->class_name());
            VERIFY_NOT_REACHED();
        }
```